### PR TITLE
[GSoC][TMVA] PyTorch Interface TMVA

### PR DESCRIPTION
**Pull request by AnirudhDagar**
This is a pull request for the development of PyTorch Interface in TMVA leveraging the PyMVA module. Feel free to comment and share feedback.

I'm also maintaining a [blog](https://anirudhdagar.github.io/gsoc/) for updating my progress and sharing some insights about my project.

# Progress

## Header
- [x] Add Initial Structure for MethodPyTorch header file.

## Class MethodPyTorch
- [x] Implement the basic outline of the MethodPyTorch Class.
- [x] Implement the DeclareOptions methods.
- [x] Implement ProcessOptions methods.
- [x] Implement SetupTorchModel method.
- [x] Implement train methods. (Done)
    - [x] Implement simple basic structure for training
    - [x] Add ability to load code from users
    - [x] Convert numpy based dataloaders in TMVA to PyTorch dataloaders
    - [x] Use default (optional optimizer SGD)
    - [x] Add savebest model
    - [x] Add scheduler
- [x] Implement Validation
- [x] Add model prediction function loading ability
- [x] Implement user code ability
- [x] Implement helper function

## Tests
- [x] Add testPyTorchClassification.C
    - [x] Add generatePyTorchModelClassification.py
    - [x] Load, Train model and test all the parameter arguments
    - [x] Verify response after model prediction function is added.
- [x] Add testPyTorchMulticlass.C
    - [x] Add generatePyTorchModelMulticlass.py
    - [x] Load, Train model and test all the parameter arguments
    - [x] Verify response after model prediction function is added.
- [x] Add testPyTorchRegression.C
    - [x] Add generatePyTorchModelRegression.py
    - [x] Load, Train model and test all the parameter arguments
    - [x] Verify response after model prediction function is added.

## GPU
- [ ] Make GPU Compatible (Require GPU for testing)

## Tutorials
These are implemented in #6273

## Documentation
- [x] Improve docs

### CODE REVIEW

**Comment by AnirudhDagar** [comment]
Please add the GSoC, WIP label to this PR for easy reference :) 
Thanks @lmoneta

**Comment by AnirudhDagar** [comment]
@lmoneta this is ready for review. Also, please add the GSoC label on this PR for convenience.
I'll be implementing the tutorials in a subsequent PR.

File: .gitignore [hunk 1]
```diff
@@ -32,4 +32,5 @@ CMakeLists.txt.user
 
 # Other Stuff
 tags
-*.d
\ No newline at end of file
+*.d
+_build/
```
Comment (ashlaban) [inline .gitignore:36]: This seems a change for your local setup. Please exclude it from the PR!

**Review by ashlaban — CHANGES_REQUESTED** [review]
Hi, thanks for providing this PR, cool to have you working for the TMVA team (I'm an old member only occasionally active).

My only comment is: Please exclude the update to the .gitignore file :)

File: .gitignore (hunk 1 above)
Comment (AnirudhDagar) [reply .gitignore]: Thanks for pointing this out. I've resolved it now :)

Comment (ashlaban) [reply .gitignore]: Excellent!

**Review by ashlaban — APPROVED** [review]
LGTM (I only did a very cursory look)

**Review by lmoneta — CHANGES_REQUESTED** [review]
There is an error in the CMake file. 
See patch fixing this: 

https://github.com/lmoneta/root/commit/2c5043ea35fedfc20ef52236c916ec2557100935

**Comment by lmoneta** [comment]
After these make fixes for the tests program, I think the PR is good to be merged.
Thank you Anirudh for this nice contribution

**Comment by AnirudhDagar** [comment]
Thanks, @lmoneta! I've applied the patch fixing cmake.

**Comment by lmoneta** [comment]
@phsft-bot build!
